File: ice0208/dp-multidimensional/1143.py
# 힌트를 이용하여 풀었습니다.
# 푼 후에 관련 영상을 시청했습니다.
# https://www.youtube.com/watch?v=ASoaQq66foQ
# https://www.youtube.com/watch?v=Ua0GhsJSlWM


class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        dp = [[0 for j in range(len(text2) + 1)] for i in range(len(text1) + 1)]

        for i in range(1, len(text1) + 1):
            for j in range(1, len(text2) + 1):
                if text1[i - 1] == text2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1] + 1
                else:
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

        return dp[len(text1)][len(text2)]


if __name__ == "__main__":
    print(Solution().longestCommonSubsequence("abcde", "ace"))
    print(Solution().longestCommonSubsequence("abc", "abc"))
    print(Solution().longestCommonSubsequence("abc", "def"))

File: moonpoison/dp_multidemensional/1143.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int longestCommonSubsequence(string t1, string t2) 
    {
        t1='0'+t1;
        t2='0'+t2;
        int dp[1001][1001];
        for(int i=0;i<t1.size();++i)
        {
            for(int j=0;j<t2.size();++j)
            {
                if(i==0||j==0)
                {
                    dp[i][j]=0;
                    continue;
                }
                if(t1[i]==t2[j]) dp[i][j]=dp[i-1][j-1]+1;
                else
                {
                    if(dp[i-1][j]>dp[i][j-1]) dp[i][j]=dp[i-1][j];
                    else dp[i][j]=dp[i][j-1];
                }
            }
        }
        return dp[t1.size()-1][t2.size()-1];
    }
};